### DISCUSSION TITLE
Trying to render in a Windows Forms panel control

### DISCUSSION THREAD

**Discussion by Anibal142857** [amerkoleci/Vortice.Windows#469] [Q&A]
Hi there, I´m working in VS2022, C#, Vortice 3.6.2, DirectX12. I was trying to have a control in Windows Forms where I can render. In order to do it I was modifying the creation of the swapchain (and the resize, etc.) using the handle and size of my control. I founded very difficult and I get errors but I know that sooner or later I´m going to have this done. The question is: Is there an easy way to do what I need that I´m not awere of? Thank you.

**Comment by amerkoleci** [comment]
Hi,
There are no easy way when talking about d3d12, you can take look at https://github.com/microsoft/DirectX-Graphics-Samples/tree/master/Samples/Desktop/D3D1211On12 and see if you can adapt it.


I'll try to setup an example if I have some spare free time

**Reply by Anibal142857** [reply]
Thank you very much for your quick response. I´m going to take a look.
The reason behind my question is that the people I´m working for asked me to determine what is the right API and wrapper in order to update their software. They are working with a very old OpenGL version and a deprecated wrapper. My test with DX11 and Vortice was successful, it covers what we need, but given the update effort, I think that is worth to try DX12 and be really up to date.   
I highly appreciate if I can get that setup example. Thank you very much again.

**Comment by MarcioAB** [comment]
Hi. This zip file is a sandbox I am working on (working in progress) for Windows Forms.
There are several examples (44 so far), each one in a separated model.
Most comments are still in my native language, but feel free to ask.

[Vortice Sandbox.zip](https://github.com/user-attachments/files/17981974/Vortice.Sandbox.zip)

**Reply by Anibal142857** [reply]
Hello Marcio, thank you for sharing your solution. I just checked it and I was able to render in a panel. It didn´t crashed. There some adjustments to make, like the mouse input area, and resize, and I can deal with it. I didn´t have time today to see the modifications in the original Vortice bindings, I´m going to do it tomorrow. I would like to keep the closest that I can to the original binding. Thank you.

**Reply by MarcioAB** [reply]
Great. At first I am using only the original Vortice, so I do not think there are any modifications related to it.

**Reply by Anibal142857** [reply]
Hello Marcio,
Now I know it is possible to do what I need. In your sandbox, I created a panel, used the handle at the D3D12_Application.InitializeD3D call. I also added the panel on the onload of the form (this.Controls.Add(pp);). These 2 modifications were enough to render at the panel.
I tried to replicate your modifications in https://github.com/amerkoleci/Vortice.Windows?tab=readme-ov-file because your model layer has no use for me, but I was not successful. I think it was too complex for my programing skills. Perhaps, you may have a suggestion for me. Thank you in advance.

**Reply by MarcioAB** [reply]
Hello Anibal, glad to know.
About "my modifications" :  what modifications?
